### ISSUE TITLE
Add support for IBM Z architecture

### ISSUE THREAD

**Issue by adamsitnik** [dotnet/BenchmarkDotNet#1711]
.NET is going to support a new architecture (https://github.com/dotnet/runtime/issues/52909), we should do so as well.

```diff
namespace System.Runtime.InteropServices
{
    public enum Architecture
    {
        X86,
        X64,
        Arm,
        Arm64,
        Wasm,
+       S390x,
    }
}
```

More details: https://github.com/dotnet/runtime/pull/52906

@uweigand some context: BenchmarkDotNet is the most popular microbenchmark harness in the .NET ecosystem. We already support x64, x86, ARM64, and ARM and would love to support `S390x` as well. When can we expect that the new architecture will be fully supported by .NET SDK and MSBuild?

**Comment by uweigand** [comment]
Hi @adamsitnik, thanks for reaching out!  We'd certainly be interested in having BenchmarkDotNet support for s390x as well.  In fact, a colleague has started to investigate what it would take to get it to run, and did run into a number of dependency issues.  I'll get back to him on that.

As to timelines, in general our goal would be to to have .NET support for s390x in place throughout the core runtime & SDK in time for the .NET 6 release.  I have most of that already running with the use of extra out-of-tree patches, and we've started the process of getting everything merged upstream -- let's see how quickly this goes.  It would be ideal if we had BenchmarkDotNet operational significantly before the release, of course, so we can already use it to evaluate and improve the performance users are going to see in the release.

**Comment by uweigand** [comment]
Thanks, that looks excellent!   We'll definitely give it a try.

**Comment by adamsitnik** [comment]
[this](https://ci.appveyor.com/project/dotnetfoundation/benchmarkdotnet/builds/39251523) CI build got green and it pushed `0.13.0.1552` NuGet to the feed

**Comment by stefan-sf-ibm** [comment]
@adamsitnik I gave `0.13.0.1552` a quick try and ran a couple of tests successfully. Thanks for the updated package!

**Comment by adamsitnik** [comment]
@stefan-sf-ibm thanks for letting me know!

**Comment by stefan-sf-ibm** [comment]
@adamsitnik I did a couple of further tests using [.NET Performance](https://github.com/dotnet/performance) and noticed that all generated `BenchmarkDotNet.Autogenerated.csproj` files contain

```
<PlatformTarget>S390x</PlatformTarget>
```

which leads to an exception [here](https://github.com/dotnet/roslyn/blob/3fa5dff8c1ffbfab5b8f6a73598cf5e3cca4ec6d/src/Compilers/Core/Portable/Compilation/Compilation.cs#L1976).

So far there exists no `System.Reflection.PortableExecutable.Machine.S390x` in [runtimelab](https://github.com/dotnet/runtimelab/blob/feature/s390x/src/libraries/System.Reflection.Metadata/src/System/Reflection/PortableExecutable/Machine.cs) and I'm not sure whether we want to add this or not. Having a look at [System.Private.CoreLib.csproj](https://github.com/dotnet/runtimelab/blob/feature/s390x/src/mono/System.Private.CoreLib/System.Private.CoreLib.csproj#L68) we have

```
<PropertyGroup Condition="'$(Platform)' == 's390x'">
  <PlatformTarget>AnyCPU</PlatformTarget>
  <DefineConstants>$(DefineConstants);TARGET_S390X;BIGENDIAN</DefineConstants>
</PropertyGroup>
```

So I'm wondering whether it would be appropriate to also set the platform target to `AnyCPU` for `S390x` for those generated `BenchmarkDotNet.Autogenerated.csproj` files?

**Comment by uweigand** [comment]
@adamsitnik I'm wondering what the benefits of setting PlatformTarget are in the first place.  According to the Roslyn docs, the only difference is that the binary is marked as limited to run only on the selected platform.  If this is true, what do we gain from using the setting here?   Or am I missing something?

**Comment by adamsitnik** [comment]
> I'm wondering what the benefits of setting PlatformTarget are in the first place.

Two things come to my mind:

- If exe is built as `AnyCpu`, on Windows it can be run both by 32bit and 64bit dotnet.exe. I guess it makes no sense for `S390x` as it is only 64bit?
- `PlatformTarget` might be taken under consideration when running dotnet restore for NuGet packages with native dependencies. 

But to be honest with you I am not 100% sure about the restore. @eerhardt could you confirm?

**Comment by uweigand** [comment]
>     * If exe is built as `AnyCpu`, on Windows it can be run both by 32bit and 64bit dotnet.exe. I guess it makes no sense for `S390x` as it is only 64bit?

Correct.  (Well, there is the legacy 32-bit linux-s390 mode, but we're not supporting that with dotnet.)

**Comment by eerhardt** [comment]
> PlatformTarget might be taken under consideration when running dotnet restore for NuGet packages with native dependencies.

I believe NuGet restore only takes into account `RuntimeIdentifier`.

https://github.com/NuGet/NuGet.Client/blob/f24bad0668193ce21a1db8cabd1ce95ba509c7f0/src/NuGet.Core/NuGet.Build.Tasks/NuGet.targets#L711-L736

**Comment by stefan-sf-ibm** [comment]
> Could you please try to extend [this](https://github.com/dotnet/performance/blob/9575b39c8278c8d841d5ab26d214439c88f3c334/src/harness/BenchmarkDotNet.Extensions/RecommendedConfig.cs#L32-L37) config definition with:
> 
> ```cs
> .WithPlatform(Environments.Platform.AnyCpu)
> ```
> 
> and see if it helps?

@adamsitnik This works out just fine, i.e., no exception anymore :)
